### PR TITLE
[transducer] first draft pr

### PR DESCRIPTION
**Pull request by Mddct**
future architecture： chunk base conformer transducer + attention rescoring
- [x] training pipeline (AED and transducer)
- [x]  model loss include: transducer loss + ctc loss(optional) + attention decoder loss (optional) 
- [  ]  python decoding： ctc greedy and prefix beam search + attention rescoring (optional)
- [  ]  cpp runtime decoding
- [  ]  more experiments
- [  ]  more predictor implementations

We need to cooperate to complete the transducer。 Suggestions welcome and hope to collaborate

### CODE REVIEW

File: wenet/transducer/test.yaml [hunk 1]
```diff
@@ -0,0 +1,107 @@
+blank_id : 0
```
Comment (robin1001) [inline wenet/transducer/test.yaml:1]: Remove the file, we can check it in related recipe.

File: wenet/transducer/main.py [hunk 2]
```diff
@@ -0,0 +1,27 @@
+import os
```
Comment (robin1001) [inline wenet/transducer/main.py:1]: It's a temp test?

File: wenet/bin/train.py [hunk 3]
```diff
@@ -42,6 +43,10 @@ def get_args():
                         default='raw',
                         choices=['raw', 'shard'],
                         help='train and cv data type')
+    parser.add_argument('--model_type',
+                        default='AED',
+                        choices=['AED', 'Transducer'],
```
Comment (robin1001) [inline wenet/bin/train.py:48]: lower case is preferred.

File: wenet/transducer/transducer.py [hunk 4]
```diff
@@ -0,0 +1,330 @@
...
+                                               text,
+                                               encoder_out_lens,
+                                               text_lengths,
+                                               blank=self.blank_id,
+                                               reduction="mean")
+        if loss_ctc is not None:
+            loss = loss + self.ctc_weight * loss_ctc.sum()
+        if loss_att is not None:
+            loss = loss + self.attention_decoder_weight * loss_att.sum()
+        return loss, loss_att, loss_ctc
```
Comment (robin1001) [inline wenet/transducer/transducer.py:130]: RNN-T loss is not retured here.

File: wenet/transducer/main.py (hunk 2 above)
Comment (Mddct) [reply wenet/transducer/main.py]: yes， will delete later

File: wenet/bin/train.py [hunk 5]
```diff
@@ -185,14 +190,18 @@ def main():
     configs['output_dim'] = vocab_size
     configs['cmvn_file'] = args.cmvn
     configs['is_json_cmvn'] = True
+    configs['blank'] = 0  # default <blank> index is zero
```
Comment (robin1001) [inline wenet/bin/train.py:193]: Just remove it since it's zero by default.

File: wenet/transducer/transducer.py [hunk 6]
```diff
@@ -0,0 +1,334 @@
...
+                                               text,
+                                               encoder_out_lens,
+                                               text_lengths,
+                                               blank=self.blank,
+                                               reduction="mean")
+        loss_rnnt = loss
+
+        # optional attention decoder
+        loss_att: Optional[torch.Tensor] = None
+        if self.transducer_weight != 1.0 and self.attention_decoder is not None:
```
Comment (robin1001) [inline wenet/transducer/transducer.py:112]: if attention_decoder_weight != 0.0

File: wenet/transducer/transducer.py [hunk 7]
```diff
@@ -0,0 +1,334 @@
...
+            loss = loss + self.ctc_weight * loss_ctc.sum()
+        if loss_att is not None:
+            loss = loss + self.attention_decoder_weight * loss_att.sum()
+        # NOTE: 'loss' must be in dict
+        return {
+            'loss': loss,
+            'loss_rnnt': loss_rnnt,
+            'loss_att': loss_att,
+            'loss_ctc': loss_ctc
+        }
```
Comment (robin1001) [inline wenet/transducer/transducer.py:134]: It's better that we keep the order compatiable with joint CTC/attention model. namely `loss`, `loss_att`, loss_ctc`, `loss_rnnt`, it's also in alphabet order.

**Review by robin1001 — APPROVED** [review]

File: wenet/utils/executor.py [hunk 8]
```diff
@@ -110,10 +109,10 @@ def train(self, model, optimizer, scheduler, data_loader, device, writer,
                     log_str = 'TRAIN Batch {}/{} loss {:.6f} '.format(
                         epoch, batch_idx,
                         loss.item() * accum_grad)
-                    if loss_att is not None:
-                        log_str += 'loss_att {:.6f} '.format(loss_att.item())
-                    if loss_ctc is not None:
-                        log_str += 'loss_ctc {:.6f} '.format(loss_ctc.item())
+                    for loss_name in loss_dict:
```
[the remark this answers is not in the document]
Comment (Mddct) [reply wenet/utils/executor.py:112]: will change later pr thx
